File: Projects/CPP-VNEXT/src/CollisionContext.cpp
#include "CollisionContext.h"

CollisionContext::CollisionContext(World& world, Organism* attacker, Organism* defender): world(world), attacker(attacker), defender(defender),
                                                                      cancelled(false), isAttackerKilled(false), isHostKilled(false), hasDefenderEvaded(false) {}

void CollisionContext::Cancel() {
    this->cancelled = true;
}

World& CollisionContext::GetWorld() const {
    return this->world;
}

Organism* CollisionContext::GetAttacker() {
    return this->attacker;
}

bool CollisionContext::IsCancelled() const {
    return this->cancelled;
}

void CollisionContext::KillAttacker() {
    this->isAttackerKilled = true;
}

void CollisionContext::KillHost() {
    this->isHostKilled = true;
}

void CollisionContext::DefenderHasEvaded() {
    this->hasDefenderEvaded = true;
}

CollisionResult CollisionContext::GetResult() const {
    if (cancelled) {
        return Cancelled;
    }

    if (this->isHostKilled && !this->isAttackerKilled) {
        return AttackerWon;
    }

    if (!this->isHostKilled && this->isAttackerKilled) {
        return DefenderWon;
    }

    if (this->hasDefenderEvaded) {
        return DefenderEvaded;
    }

    if (!this->isHostKilled) {
        if (attacker->GetStrength() >= defender->GetStrength()) {
            return AttackerWon;
        }else {
            return DefenderWon;
        }
    }

    return BothDied;
}

File: Projects/Java/src/main/java/com/nieradko/worldsim/core/CollisionContext.java
package com.nieradko.worldsim.core;

import com.nieradko.worldsim.core.animals.Animal;

import java.util.Optional;

public class CollisionContext implements ICollisionContext {
    private final Animal attacker;
    private final Organism defender;
    private boolean isCancelled = false;
    private boolean hasDefenderEvaded = false;
    private boolean hasDefenderDied = false;
    private boolean hasAttackerDied = false;

    public CollisionContext(Animal attacker, Organism defender) {
        this.attacker = attacker;
        this.defender = defender;
    }

    @Override
    public Animal getAttacker() {
        return attacker;
    }

    @Override
    public Organism getDefender() {
        return defender;
    }

    @Override
    public void cancel() {
        isCancelled = true;
    }

    @Override
    public void defenderHasEvaded() {
        hasDefenderEvaded = true;
    }

    @Override
    public void defenderHasDied() {
        hasDefenderDied = true;
    }

    @Override
    public void attackerHasDied() {
        hasAttackerDied = true;
    }

    @Override
    public boolean isResolved() {
        return isCancelled || hasDefenderEvaded || hasDefenderDied || hasAttackerDied;
    }

    @Override
    public Optional<CollisionResult> getResult() {
        if (!isResolved()) {
            return Optional.empty();
        }

        if (isCancelled) {
            return Optional.of(CollisionResult.Cancelled);
        }

        if (hasDefenderDied && !hasAttackerDied) {
            return Optional.of(CollisionResult.AttackerWon);
        }

        if (!hasDefenderDied && hasAttackerDied) {
            return Optional.of(CollisionResult.DefenderWon);
        }

        if (hasDefenderDied) {
            return Optional.of(CollisionResult.BothDied);
        }

        if (attacker.getStrength() >= defender.getStrength()) {
            return Optional.of(CollisionResult.AttackerWon);
        }else {
            return Optional.of(CollisionResult.DefenderWon);
        }
    }
}
